specification/keyvault/data-plane/Microsoft.KeyVault/stable/7.3/certificates.json

```diff
@@ -1630,7 +1630,7 @@
             "AutoRenew"
           ],
           "x-ms-enum": {
-            "name": "ActionType",
+            "name": "CertificateActionType",
```

Comment:
This is a breaking change. Why is this necessary, though? This name only affects SDKs, which should be referencing individual swaggers to generate separate SDKs for keys, secrets, certificates, and administration (basically, everything else). So conflicts between the two are irrelevant.

Reply:
`$.definitions.Actions` should be `$.definitions.Action`

```
  - from: certificates.json
    where: $.definitions.Action
    transform: $.properties.action_type["x-ms-enum"].name = "CertificateActionType";
  - from: keys.json
    where: $.definitions.LifetimeActionsType
    transform: $.properties.type["x-ms-enum"].name = "LifetimeActionType";
```

Reply:
@heaths thanks for the suggestion.

> This is a breaking change for SDKs but shouldn't be necessary. This name only affects SDKs, which should be referencing individual swaggers to generate separate SDKs for keys, secrets, certificates, and administration (basically, everything else). So conflicts between the two are irrelevant unless you're generating a single SDK targeting all swaggers, though.

From our side we're trying to generate the SDK using the same approach that Track1 is using today, namely keeping this in a single package - so I think the autorest transform is _probably_ fine as a workaround here, but it begs the question of why not fix the data rather than applying a workaround?

From a logical deployment perspective, since these operations are exposed in the same Swagger directory, I'd expect them to be deployed together - so I'd argue that if these are intended to be separate SDKs then these should be split out (which would also remove the need for this workaround). Whilst this may seem like a pedantic point, fixing the data (either by merging this PR as-is, or by splitting these files into different directories) seems preferable to applying autorest workarounds, which should presumably only be a short-term solution and not widely used, else the source data isn't accurate?

> That begs the question, though: have you considered using our Go SDKs? https://github.com/Azure/azure-sdk-for-go/tree/main/sdk/keyvault

Unfortunately we're unable to use the Track2 SDK's for reasons we've raised with the Azure SDK team previously - instead [we're using `hashicorp/go-azure-sdk`](https://github.com/hashicorp/go-azure-sdk) which is generated from [`hashicorp/pandora`](https://github.com/hashicorp/pandora) (for Resource Manager) - and we'll be generating some Data Plane packages into another repository until we look into importing/generating those.

Reply:
There are a lot of services that don't use swagger in non-standard ways. Going forward with Cadl - a new description language that can generate both server- and client-side code (and swaggers, for legacy support) we'll create subdirectories per SDK, so these files won't be in the same directory. Ever since we refactored them, they were never meant to represent a single service.

Why can't you use the track 2? I've talked with several teams in the past and showed them our migration guides e.g., https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/keyvault/Azure.Security.KeyVault.Secrets/MigrationGuide.md. Everything should be supported 1:1 between track 1 and track 2 SDKs already except for SAS (not recommended anyway, which is why we didn't generate an SDK).

Reply:
@heaths

>  It was a lot of extra work for almost every SDK language to trim support for endpoints that weren't supposed to be in an SDK, like keys and certs endpoints in the secrets SDK.

Potentially an odd take, but why would these be stripped out? If the Swagger documents them the SDK's should be generating them.

> Going forward with Cadl - a new description language that can generate both server- and client-side code (and swaggers, for legacy support) we'll create subdirectories per SDK, so these files won't be in the same directory. Ever since we refactored them, they were never meant to represent a single service.

Do you have any more context on Cadl / the timeframe there?

> Why can't you use the track 2?

We've previously given copious amounts of feedback to the Azure SDK Team as to why we're unable to use the Track2 Go SDK - and since we didn't hear back from the Azure SDK Team about our concerns - ultimately ended up having to build [`hashicorp/go-azure-sdk`](https://github.com/hashicorp/go-azure-sdk).

[The `hashicorp/go-azure-sdk` repository](https://github.com/hashicorp/go-azure-sdk) both fixes the concerns we had (albeit using the Track1 base layer as a temporary stop-gap, until we're fully off the Azure SDK for Go) and enables us to embed logic within the Go SDK to both make it more consistent (for example, [all operations take a Resource ID object](https://pkg.go.dev/github.com/hashicorp/go-azure-sdk@v0.20221024.1202112/resource-manager/compute/2021-11-01/virtualmachines#VirtualMachinesClient) where a Resource is being mutated) - which ultimately reduces the amount of code needed to interact with the SDK.

By comparison, adopting Track2 would mean:

* increasing the amount of code needed (both per-operation and in general)
* more admin overhead with nested Go modules (e.g. which semver relates to which API version? a large number of breaking changes every time we update a module)
* re-introducing the pain of composite Swaggers (which comes back to more admin overhead/which API version is _actually_ being used)
* hitting up against more Swagger inconsistencies (since `hashicorp/go-azure-sdk` corrects a bunch of these)

As such whilst Track2 may work great for folks using a small number of packages, unfortunately the design/architectural decisions mean it doesn't work for us since we're vendoring large amounts of the SDK (and are concerned about which API versions we're using) - as such this isn't something we're planning to adopt (even if the amount of code required to use Track2 was reduced in the same manner as `hashicorp/go-azure-sdk`).

Whilst that's unfortunate, in practice not using the Track2 SDK has enabled us to look into additional functionality (for example, Terraform Resource Generation) that wouldn't have been possible if we used the Track1 or Track2 SDKs (since we're normalizing the Swaggers, for example duck-typing into common types) - so whilst there is an overhead to generating this SDK ourselves, I'd argue it's working out really well for us - and as such unfortunately we have no plans to adopt Track2.

Reply:
The secrets package doesn't need support for keys or certificates endpoints, for example. We split packages up to be more area-focused, so having one giant file with all the endpoints didn't make sense. There are also numerous feature teams and refactoring the files made it easier for them to work and deploy in parallel as well.

We don't have a timeframe for when Key Vault will be migrated to Cadl, but you can find more information about Cadl here: https://github.com/microsoft/cadl

As of now, however, these swaggers have shipped this way for years and changing now is a breaking change. Please use the suggested (amended) autorest transforms.

Reply:
@heaths 

Just to be clear, could you highlight _where_ this is a breaking change? Splitting this out is a breaking change, since the Azure SDK for Go has shipped this as a combined Swagger for many years:

* Create Certificate - https://github.com/Azure/azure-sdk-for-go/blob/438bb470af0f10f6952c59605ad8d7ef86acffba/services/keyvault/v7.1/keyvault/client.go#L367

* Create Key - https://github.com/Azure/azure-sdk-for-go/blob/438bb470af0f10f6952c59605ad8d7ef86acffba/services/keyvault/v7.1/keyvault/client.go#L462

* Set Secret - https://github.com/Azure/azure-sdk-for-go/blob/438bb470af0f10f6952c59605ad8d7ef86acffba/services/keyvault/v7.1/keyvault/client.go#L6092

As such I'm confused as to how this could be considered a breaking change? I can understand the desire to split this, fine, but I don't believe this is a breaking change.

Whilst we can look to add an AutoRest Transform to workaround this, my understanding is those are supposed to be a last resort - so I'm at a loss for why:

a) applying this change would be a breaking change (to either Track1, or Track2)
b) the recommended approach here is to apply a hack (an AutoRest Transform) rather than fix the underlying issue

Reply:
any update @heaths?

Reply:
Fundamentally the differentiation between Track1 and Track2 is irrelevant here though, this change wouldn't be a breaking change in either of them, so the reason for closing here is pretty misleading?

Reply:
@heaths I get that you'd like us to use Track2, but I still think you've closed this for the wrong reasons here - @JeffreyRichter any chance you'd be able to take a look here? I think @heaths and I are at an impasse here?

Reply:
I'll meet with Heath and discuss so I understand the issue and report back here (hopefully next week but it is Thanksgiving).

Reply:
@JeffreyRichter did you have a chance to chat with Heath about this yet?

Reply:
Yes, Heath & I spoke and as it is SDK breaking, we cannot accept this PR.

Reply:
Any chance you could explain *how* this is SDK breaking? The Breaking Change Action above says this isn't a breaking change:

<img width="586" alt="Screenshot 2022-12-12 at 10 01 44" src="https://user-images.githubusercontent.com/666005/207004245-19703be3-3829-4fe8-9151-ed5dc02d92e0.png">